cpp/src/arrow/compute/kernels/match_test.cc

```diff
@@ -0,0 +1,457 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <algorithm>
+#include <cstdint>
+#include <cstdio>
+#include <functional>
+#include <locale>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "arrow/array.h"
+#include "arrow/buffer.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/match.h"
+#include "arrow/compute/kernels/util_internal.h"
+#include "arrow/compute/test_util.h"
+#include "arrow/memory_pool.h"
+#include "arrow/status.h"
+#include "arrow/table.h"
+#include "arrow/testing/gtest_common.h"
+#include "arrow/testing/util.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/decimal.h"
+
+namespace arrow {
+namespace compute {
+
+// ----------------------------------------------------------------------
+// Match tests
+
+template <typename Type, typename T = typename TypeTraits<Type>::c_type>
+void CheckMatch(FunctionContext* ctx, const std::shared_ptr<DataType>& type,
+                const std::vector<T>& in_values, const std::vector<bool>& in_is_valid,
+                const std::vector<T>& member_set_values,
+                const std::vector<bool>& member_set_is_valid,
+                const std::vector<int32_t>& out_values,
+                const std::vector<bool>& out_is_valid) {
+  std::shared_ptr<Array> input = _MakeArray<Type, T>(type, in_values, in_is_valid);
+  std::shared_ptr<Array> member_set =
+      _MakeArray<Type, T>(type, member_set_values, member_set_is_valid);
+  std::shared_ptr<Array> expected =
+      _MakeArray<Int32Type, int32_t>(int32(), out_values, out_is_valid);
+
+  Datum datum_out;
+  ASSERT_OK(Match(ctx, input, member_set, &datum_out));
+  std::shared_ptr<Array> result = datum_out.make_array();
+  ASSERT_ARRAYS_EQUAL(*expected, *result);
+}
+
+class TestMatchKernel : public ComputeFixture, public TestBase {};
+
+template <typename Type>
+class TestMatchKernelPrimitive : public ComputeFixture, public TestBase {};
+
+typedef ::testing::Types<Int8Type, UInt8Type, Int16Type, UInt16Type, Int32Type,
+                         UInt32Type, Int64Type, UInt64Type, FloatType, DoubleType,
+                         Date32Type, Date64Type>
+    PrimitiveDictionaries;
+
+TYPED_TEST_CASE(TestMatchKernelPrimitive, PrimitiveDictionaries);
+
+TYPED_TEST(TestMatchKernelPrimitive, Match) {
+  using T = typename TypeParam::c_type;
+  auto type = TypeTraits<TypeParam>::type_singleton();
+
+  // No Nulls
+  CheckMatch<TypeParam, T>(&this->ctx_, type,
+                           /* in_values= */ {2, 1, 2, 1, 2, 3},
+                           /* in_is_valid= */ {},
+                           /* member_set_values= */ {2, 1, 2, 3},
+                           /* member_set_is_valid= */ {},
+                           /* out_values= */ {0, 1, 0, 1, 0, 2},
```

Comment:
I would have expected the last entry to be 3, since the index of "3" in `member_set_values is 3. The doccomment for Match should mention that the member set is uniqued first or something